src/mmds/src/token.rs

```diff
@@ -0,0 +1,562 @@
+            warn!(
+                "MMDS's token authority entity has been reseeded. \
+                All previously created tokens are now invalid."
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Validate the token time to live against bounds.
+    fn check_ttl(ttl_seconds: u32) -> bool {
+        MIN_TOKEN_TTL_SECONDS <= ttl_seconds && ttl_seconds <= MAX_TOKEN_TTL_SECONDS
+    }
+
+    /// Compute expiry time in seconds by adding the time to live provided
+    /// to the current time measured in seconds.
+    fn compute_expiry(ttl_seconds: u32) -> u64 {
+        // Get time elapsed since UNIX_EPOCH.
+        let now_as_duration = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("System time is before UNIX_EPOCH.");
+        // Convert elapsed time into seconds.
+        let ttl_duration = Duration::from_secs(u64::from(ttl_seconds));
+
+        // Compute expiry by adding ttl value to current time in seconds.
+        // This addition is safe because ttl is verified beforehand and
+        // can never be more than 6h.
+        let expiry_duration = now_as_duration.add(ttl_duration);
+
+        expiry_duration.as_secs()
+    }
+}
+
+/// Structure for token information.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+struct Token {
+    version: u16,
+    // Nonce or Initialization Vector.
+    iv: [u8; IV_LEN],
+    // Encrypted expire time.
+    payload: [u8; PAYLOAD_LEN],
+    // Tag returned after encryption.
+    tag: [u8; TAG_LEN],
+}
+
+impl Token {
+    /// Create a new token struct.
+    fn new(iv: [u8; IV_LEN], payload: [u8; PAYLOAD_LEN], tag: [u8; TAG_LEN]) -> Self {
+        Token {
+            version: 1,
+            iv,
+            payload,
+            tag,
+        }
+    }
+
+    /// Encode token structure into a string using base64 encoding.
+    fn base64_encode_token(&self) -> Result<String, Error> {
+        let token_bytes: Vec<u8> = DefaultOptions::new()
+            .with_fixint_encoding()
+            .allow_trailing_bytes()
+            .with_limit(DESERIALIZATION_BYTES_LIMIT)
+            .serialize(self)
+            .map_err(|_| Error::TokenGeneration)?;
+
+        // Encode token structure bytes into base64.
+        Ok(base64::encode_config(token_bytes, base64::STANDARD))
+    }
+
+    /// Decode token structure from base64 string.
+    fn base64_decode_token(encoded_token: &str) -> Result<Self, Error> {
+        let token_bytes = base64::decode_config(encoded_token, base64::STANDARD)
+            .map_err(|_| Error::ExpiryExtraction)?;
+
+        let token: Token = DefaultOptions::new()
+            .with_fixint_encoding()
+            .allow_trailing_bytes()
+            .with_limit(DESERIALIZATION_BYTES_LIMIT)
+            .deserialize(&token_bytes)
+            .map_err(|_| Error::ExpiryExtraction)?;
+        Ok(token)
+    }
+
+    /// Validate constant fields inside the token structure.
+    fn sanity_check(&self) -> bool {
```

Comment:
The sanity check is not used on the token creation path, but on the validation path. Since the token is created by MMDS itself, and we explicitly set the `version` field as 1, there is no need to verify this during token creation. 

The check is, however, required on the token validation path: when responding to the `GET` request, after decoding the `Token` struct from the base64 string, the `version` field is verified. 

The `sanity_check` function is not public and is only used by the validation function: `is_valid()`.

Reply:
I see. It can be moved to `base64_decode_token` then. But it's fine keeping it here as well